### PR TITLE
Rails 5.1 support

### PR DESCRIPTION
**Pull request by bricesanchez**
Rails 5.1 support is under progress.

## TODO (updated 2017/08/11)
- [x] Add support for Rails 5.1
- [x] Fix SEO Meta, wait for PR merge: https://github.com/parndt/seo_meta/pull/31
- [x] Review of @parndt

### CODE REVIEW

**Comment by parndt** [comment]
The private routes reader is a bit of a pain. Is there another way to inspect the routes?  Do we still need to do this?

**Comment by deepj** [comment]
@bricesanchez I'm getting these errors when I run specs https://gist.github.com/deepj/9b4c2762d824e9bdd1b3

**Comment by bricesanchez** [comment]
@deepj What i see is not so bad, there is a deprecation warning but your failling specs are only on images manipulations. So i guess you don't have image magick installed on your computer!?

**Comment by bricesanchez** [comment]
@parndt I've no idea at this time.

**Comment by bricesanchez** [comment]
It moves forward and well!

I've discovered that i've to update `refinerycms-wymeditor` in order to be able to fix some specs.

We are close to be able to make a Q&A.

```ruby
Finished in 2 minutes 26.1 seconds (files took 11.64 seconds to load)
532 examples, 12 failures, 8 pending

Failed examples:

rspec ./images/spec/features/refinery/admin/images_spec.rb:67 # Page Visual Editor - Add Image When there are many images has behaviour: indexes images shows all the images
rspec './images/spec/features/refinery/admin/images_spec.rb[2:3:3:1:1]' # Page Visual Editor - Add Image When there are many images has behaviour: uploads images when the image type is acceptable the image is uploaded
rspec './images/spec/features/refinery/admin/images_spec.rb[2:3:3:2:1]' # Page Visual Editor - Add Image When there are many images has behaviour: uploads images when the image type is not acceptable the image is rejected
rspec './images/spec/features/refinery/admin/images_spec.rb[2:1:1:1:1]' # Page Visual Editor - Add Image When there are no images has behaviour: uploads images when the image type is acceptable the image is uploaded
rspec './images/spec/features/refinery/admin/images_spec.rb[2:1:1:2:1]' # Page Visual Editor - Add Image When there are no images has behaviour: uploads images when the image type is not acceptable the image is rejected
rspec ./images/spec/features/refinery/admin/images_spec.rb:58 # Page Visual Editor - Add Image When there is one image has behaviour: indexes images shows all the images
rspec './images/spec/features/refinery/admin/images_spec.rb[2:2:3:2:1]' # Page Visual Editor - Add Image When there is one image has behaviour: uploads images when the image type is not acceptable the image is rejected
rspec './images/spec/features/refinery/admin/images_spec.rb[2:2:3:1:1]' # Page Visual Editor - Add Image When there is one image has behaviour: uploads images when the image type is acceptable the image is uploaded
rspec ./core/spec/lib/refinery/crud_spec.rb:113 # Refinery::CrudDummyController#update_positions sorts numerically rather than by string key
rspec ./pages/spec/presenters/refinery/pages/menu_presenter_spec.rb:66 # Refinery::Pages::MenuPresenter#to_html returns menu items wrapped in html
rspec ./pages/spec/presenters/refinery/pages/menu_presenter_spec.rb:75 # Refinery::Pages::MenuPresenter#to_html takes mount point into account for normal pages
rspec ./pages/spec/presenters/refinery/pages/menu_presenter_spec.rb:85 # Refinery::Pages::MenuPresenter#to_html takes mount point into account when page has a link_url the menu item URL includes the mounted path
```

**Comment by henry-p** [comment]
Hello Brice, great to see the progress here! Is it possible for you to give a time estimate for the completion of the Rails 5 support? Cheers, Henry

**Comment by bricesanchez** [comment]
@henry-p ASAP or When it will be ready ... hahaha

No joke, i'm stuck with an assets error, i've done most of changes to be close to release a first beta.

I've probably need the help of @parndt to find a solution.

Don't hesitate to test this branch and return bugs :)

**Comment by digitalextremist** [comment]
My development team is adopting `Refinery CMS` and is willing to contribute time to `Rails 5` support. It seems like it has been some time since this PR was touched. Does anyone mind giving an update/recap of where the overall project stands?

**Comment by bricesanchez** [comment]
I've found that :set_routes_reloader_hook is probably the guilty for the assets problem: https://github.com/rails/sprockets-rails/pull/388#issuecomment-312468797

File: core/lib/refinery/crud.rb [hunk 1]
```diff
@@ -283,7 +283,8 @@ def reorder
             # Based upon https://github.com/matenia/jQuery-Awesome-Nested-Set-Drag-and-Drop
             def update_positions
               previous = nil
-              params[:ul].each do |_, list|
+              # TODO: Find fix because `to_unsafe_h` badly reorder the hash
```
Comment (bricesanchez) [inline core/lib/refinery/crud.rb]: @sjoulbak i don't know if it's still an up-to-date bug but it's probably why the following specs failed: https://github.com/refinery/refinerycms/pull/3122/files#diff-b2b23769a8e7837f008b49da20475dc5R125

Comment (sjoulbak) [inline core/lib/refinery/crud.rb]: Hmm I'll investigate it this morning

Comment (sjoulbak) [inline core/lib/refinery/crud.rb]: Removing the `to_unsafe_h` function here doesn't effect the tests. We'll only get a deprecation warning.

File: core/refinerycms-core.gemspec [hunk 2]
```diff
@@ -27,8 +27,9 @@ Gem::Specification.new do |s|
   s.add_dependency 'railties',                    rails_version
   s.add_dependency 'activerecord',                rails_version
   s.add_dependency 'actionpack',                  rails_version
+  s.add_dependency 'actionmailer',                rails_version
```
Comment (parndt) [inline core/refinerycms-core.gemspec]: why do we need to depend on actionmailer now?

File: images/refinerycms-images.gemspec [hunk 3]
```diff
@@ -20,7 +20,8 @@ Gem::Specification.new do |s|
   s.test_files        = `git ls-files -- spec/*`.split("\n")
 
   s.add_dependency 'dragonfly',               '~> 1.0.0'
-  s.add_dependency 'globalize',               ['>= 4.0.0', '< 5.2']
+  s.add_dependency 'globalize',               '5.1.0.beta1'
```
Comment (parndt) [inline images/refinerycms-images.gemspec]: this should probably specify the beta as the minimum, i.e. `['~> 5.1.0', '>= 5.1.0.beta1']`?

File: pages/lib/refinery/pages/engine.rb [hunk 4]
```diff
@@ -25,11 +25,13 @@ class Engine < ::Rails::Engine
...
       end
 
-      initializer "refinery.pages append marketable routes", :after => :set_routes_reloader_hook do
+      # FIXME: Find why in Rails 5, if we use `:set_routes_reloader_hook` it redirect sprockets assets engines paths to 404
+      initializer "refinery.pages append marketable routes"  do
         append_marketable_routes if Refinery::Pages.marketable_urls
       end
 
-      initializer "add marketable route parts to reserved words", :after => :set_routes_reloader_hook do
+      # FIXME: Find why in Rails 5, if we use `:set_routes_reloader_hook` it redirect sprockets assets engines paths to 404
```
Comment (parndt) [inline pages/lib/refinery/pages/engine.rb]: FIXME

File: pages/lib/refinery/pages/engine.rb [hunk 5]
```diff
@@ -25,11 +25,13 @@ class Engine < ::Rails::Engine
         Refinery.include_once(Refinery::AdminController, Refinery::Pages::Admin::InstanceMethods)
       end
 
-      initializer "refinery.pages append marketable routes", :after => :set_routes_reloader_hook do
+      # FIXME: Find why in Rails 5, if we use `:set_routes_reloader_hook` it redirect sprockets assets engines paths to 404
```
Comment (parndt) [inline pages/lib/refinery/pages/engine.rb]: FIXME

File: pages/lib/refinery/pages/engine.rb [hunk 6]
```diff
@@ -49,14 +52,13 @@ def append_marketable_routes
...
         end
-        Rails.application.routes_reloader.reload!
       end
 
       # Add any parts of routes as reserved words.
       def add_route_parts_as_reserved_words
         ActiveSupport.on_load(:active_record) do
           # do not add routes with :allow_slug => true
-          included_routes = Rails.application.routes.named_routes.routes.reject{ |name, route| route.defaults[:allow_slug] }
+          included_routes = Rails.application.routes.named_routes.to_a.reject{ |name, route| route.defaults[:allow_slug] }
```
Comment (parndt) [inline pages/lib/refinery/pages/engine.rb]: I understand why the `.to_a` was added, but how come the `.routes` part was dropped?

File: pages/refinerycms-pages.gemspec [hunk 7]
```diff
@@ -20,8 +20,9 @@ Gem::Specification.new do |s|
   s.test_files        = `git ls-files -- spec/*`.split("\n")
 
   s.add_dependency 'friendly_id',                 ['>= 5.1.0', '< 5.3']
-  s.add_dependency 'globalize',                   ['>= 4.0.0', '< 5.2']
-  s.add_dependency 'awesome_nested_set',          '~> 3.0.0'
+  s.add_dependency 'globalize',                   '5.1.0.beta1'
```
Comment (parndt) [inline pages/refinerycms-pages.gemspec]: same comment about minimum version number as above with the images gemspec.

File: resources/refinerycms-resources.gemspec [hunk 8]
```diff
@@ -21,7 +21,8 @@ Gem::Specification.new do |s|
 
   s.add_dependency 'acts_as_indexed',         '~> 0.8.0'
   s.add_dependency 'dragonfly',               '~> 1.0.0'
-  s.add_dependency 'globalize',               ['>= 4.0.0', '< 5.2']
+  s.add_dependency 'globalize',               '5.1.0.beta1'
```
Comment (parndt) [inline resources/refinerycms-resources.gemspec]: same comment about minimum version number as above with the images gemspec.

File: core/app/assets/stylesheets/config/_backend.scss [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+//   #page-preview, .frontend-icons {
+//     .icon {
+//       font-family: "GcSite" !important;
+//     }
+//   }
+// }
+//
+// body .icon.icon-fe {
+//   font-family: "GcSite" !important;
+// }
```
Comment (bricesanchez) [inline core/app/assets/stylesheets/config/_backend.scss]: Nope, i will remove it

File: Gemfile [hunk 10]
```diff
@@ -9,16 +9,31 @@ path "./" do
...
+# Rails 5 temp. gem sources
+
+gem 'refinerycms-i18n', git: 'https://github.com/refinery/refinerycms-i18n', branch: 'feature/rails-5'
+
+group :development do
+  gem 'listen', '~> 3.0'
+end
+
+#
+###########################
```
Comment (bricesanchez) [inline Gemfile]: I will check that. But until we don't merge this PR https://github.com/refinery/refinerycms-i18n/pull/65 with a version bump, i will have to specify this source

File: core/app/assets/stylesheets/config/_variables.scss [hunk 11]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Define site wide variables here, like primary color & font
+ *
+ ******************************************************************************
+ * For full documentation of what is being included where, see:
+ * refinerycms gem: /core/app/assets/stylesheets/refinery-glass.scss
```
Comment (bricesanchez) [inline core/app/assets/stylesheets/config/_variables.scss]: Nope

File: core/lib/refinery.rb [hunk 12]
```diff
@@ -3,6 +3,8 @@
 require 'action_controller'
 require 'rbconfig'
 require 'font-awesome-sass'
+require 'jquery-rails'
+require 'jquery-ui-rails'
```
Comment (bricesanchez) [inline core/lib/refinery.rb]: It's a dependency of core : https://github.com/refinery/refinerycms/blob/dc6aa979600405613f0d5a181e92cf3eb4f0e657/core/refinerycms-core.gemspec#L35

I will check if we still need to require it here. A long time ago, it fixed a bug.

File: core/lib/refinery/core/engine.rb [hunk 13]
```diff
@@ -63,7 +63,8 @@ def refinery_inclusion!
         end
       end
 
-      initializer "refinery.routes", :after => :set_routes_reloader_hook do |app|
+      # FIXME: Find why in Rails 5, if we use `:set_routes_reloader_hook` it redirect sprockets assets engines paths to 404
```
Comment (bricesanchez) [inline core/lib/refinery/core/engine.rb]: I need your expertise to understand the changes between Rails 4 and 5. But in all cases the changes, fix our problem

File: core/lib/refinery/crud.rb (hunk 1 above)
Comment (bricesanchez) [inline core/lib/refinery/crud.rb]: I will check that

File: images/refinerycms-images.gemspec (hunk 3 above)
Comment (bricesanchez) [inline images/refinerycms-images.gemspec]: yes it will be better

File: core/refinerycms-core.gemspec (hunk 2 above)
Comment (bricesanchez) [inline core/refinerycms-core.gemspec]: It was required by Rails 5.0.0.beta3, let me check if it's still relevant: https://github.com/refinery/refinerycms/commit/9219ecf5d56f3a2e8b08ec3dfe8528e9d92f30ca

File: pages/lib/refinery/pages/engine.rb (hunk 6 above)
Comment (bricesanchez) [inline pages/lib/refinery/pages/engine.rb]: It comes from this: https://github.com/rails/rails/commit/7187339854d8454fdb3338b3041f89c6ce7e1e41

Comment (parndt) [inline pages/lib/refinery/pages/engine.rb]: Thanks

File: core/refinerycms-core.gemspec (hunk 2 above)
Comment (parndt) [inline core/refinerycms-core.gemspec]: I hope it's not needed!

File: Gemfile (hunk 10 above)
Comment (parndt) [inline Gemfile]: Happy to merge that.

**Comment by bricesanchez** [comment]
@parndt I'm ready for the 2nd pass of review

File: core/refinerycms-core.gemspec [hunk 14]
```diff
@@ -2,7 +2,7 @@
 require File.expand_path('../../core/lib/refinery/version', __FILE__)
 
 version = Refinery::Version.to_s
-rails_version = ['>= 4.2.3', '< 5.0']
+rails_version = ['>= 5.0.0', '< 5.1']
```
Comment (parndt) [inline core/refinerycms-core.gemspec]: Does it not work with Rails 5.1.2? Could we relax this to find out?

Comment (bricesanchez) [inline core/refinerycms-core.gemspec]: At this time no.
I've send you some PRs/Issues on Globalize, Decoratos and Seo_meta gems. 
After we have 37 tests to fix, nothing too hard IMHO. 

Should we start using Appraisals?

Comment (parndt) [inline core/refinerycms-core.gemspec]: I've merged your PRs and pushed new gem versions 😄

Comment (parndt) [inline core/refinerycms-core.gemspec]: If you want to do the work to get Appraisals working then by all means give it a go!

Comment (bricesanchez) [inline core/refinerycms-core.gemspec]: After thinking about it, we should only support the latest version of Rails, not all versions. So I do not recommend to use Appraisals.

**Comment by bricesanchez** [comment]
Specs will succeed when seo_meta will be patched: https://github.com/parndt/seo_meta/pull/31

**Comment by bricesanchez** [comment]
It's ready to be reviewed @parndt

**Comment by parndt** [comment]
Thanks for all the hard work!

**Comment by bricesanchez** [comment]
I'm so happy it's merged!

**Comment by sjoulbak** [comment]
Well done!

**Comment by bsodmike** [comment]
Fabulous work all, pleasure to be part of it!

**Comment by bsodmike** [comment]
Trust the README will be revised soon?

**Comment by bricesanchez** [comment]
Yes i will revise the README/Changelog at the release.

**Comment by parndt** [comment]
I left the branch not deleted in case someone's using it in their Gemfile. I'll try to clean it up in the coming weeks though.

**Pull request merged**